### PR TITLE
Spark: Remove backup table after a successful migrate action.

### PR DESCRIPTION
**Pull request by sririshindra**
Once the MigrateTable action is called, a backup table is created. The backup table is used to restore the original table in case the MigrateTable action fails. This restoration is done by renaming the backup table to the original table. Therefore if the migrate operation fails there is no backup table lying around in the catalog. However, if the migration goes through successfully, the backup table is not cleaned up leaving the user to find a new unknown table in their catalog.

This PR cleans up the backup table after the migrate table action succeeds.

I tested this manually and I also added additional assertions to the existing unit tests to verify that the backup table is cleaned up properly.

### CODE REVIEW

**Comment by sririshindra** [comment]
@aokolnychyi , @rdblue , @danielcweeks 
Could you please take a look at this PR if you get a chance? Thanks.

**Comment by szehon-ho** [comment]
I did notice it the other day, and also thought it would be a good usability feature (especially as there is snapshot table that keeps the original table).  I wonder if need to have a  drop_backup flag or something, in case someone depends on original behavior somehow?  Let's see if everyone is ok with it.

File: spark/v3.1/spark/src/main/java/org/apache/iceberg/spark/actions/BaseMigrateTableSparkAction.java
Comment (szehon-ho) [inline spark/v3.1/spark/src/main/java/org/apache/iceberg/spark/actions/BaseMigrateTableSparkAction.java:228]: Should we do try/catch?

**Comment by wypoon** [comment]
@RussellSpitzer were you the original author of the migrate action (#1525)? Was the purpose of the backup table simply for restoring it in case the migrate failed? If so, it would make sense to remove the backup if the migrate succeeded, right?

**Comment by rdblue** [comment]
I'm -1 on this. Not removing the backup table is intentional so that the user could choose which table to move forward with. Our docs included a step to "remove the backup table" once everything was running smoothly.

I'm fine adding this as an option, but I don't think it should be the default.

**Comment by wypoon** [comment]
@rdblue can you please point me to the doc with a step to remove the backup table? I do not see it in https://iceberg.apache.org/docs/latest/spark-procedures/#migrate.

**Comment by sririshindra** [comment]
> 

@rdblue In that case how would you feel about letting the user decide on the name of the backup table instead of the generic backup suffix appended to the table name? I think we should at least document this behavior so that users are not surprised to see a new table in their catalog that they did not create. I can also do a documentation update if that is what is more appropriate.

**Comment by szehon-ho** [comment]
Docs are fine, but to me it would make more sense to add a non-default flag to drop_backup (as mentioned above), if we want to do add a feature here.   I dont see letting the user decide name for backup is that helpful?

**Comment by sririshindra** [comment]
> Docs are fine, but to me it would make more sense to add a non-default flag to drop_backup (as mentioned above), if we want to do add a feature here. I dont see letting the user decide name for backup is that helpful?

Sounds good. @szehon-ho I will update the PR to add the drop_table flag and make it optional. Thanks.

**Comment by wypoon** [comment]
I understand the thinking behind keeping the backup table after a successful migrate action and having the user decide what to do with it. If somehow this isn't already documented, then @sririshindra you could open a doc PR to document it.
I think giving the user an option to remove the backup table automatically upon a successful migrate action is still useful. We can have a boolean parameter, defaulting to false, for removing the backup table, and pass it in the migrate procedure.
I'm neutral on having an option for naming the backup table.

File: api/src/main/java/org/apache/iceberg/actions/ActionsProvider.java
Comment (szehon-ho) [inline api/src/main/java/org/apache/iceberg/actions/ActionsProvider.java:38]: We should just keep this the same, and use the options already existing on MigrateTables.

File: spark/v3.0/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMigrateTableProcedure.java
Comment (szehon-ho) [inline spark/v3.0/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMigrateTableProcedure.java:87]: I thought it's backward compatible, why do we need to change existing tests?

File: api/src/main/java/org/apache/iceberg/actions/ActionsProvider.java
Comment (sririshindra) [reply api/src/main/java/org/apache/iceberg/actions/ActionsProvider.java]: Done, Fixed in the latest commit.

File: spark/v3.0/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMigrateTableProcedure.java
Comment (sririshindra) [reply spark/v3.0/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMigrateTableProcedure.java]: Fixed in the latest commit.

File: spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/actions/BaseMigrateTableSparkAction.java
Comment (szehon-ho) [inline spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/actions/BaseMigrateTableSparkAction.java:116]: So additionalProperties() looks like properties that will be set on the target table, right?  That's not great because then there will be this extra property 'dropBackup' that's useless for the Iceberg table.

So I think , let's add an extra method on MigrateTableSparkAction.dropBackup(boolean), and propagate it via its own dedicated argument in the procedure?

File: api/src/main/java/org/apache/iceberg/actions/ActionsProvider.java
Comment (szehon-ho) [reply api/src/main/java/org/apache/iceberg/actions/ActionsProvider.java]: Better but not exactly what I meant, I think we still need an extra method on MigrateTableAction object (MigrateTableAction::dropBackup(boolean)).  Otherwise piggybacking with properties there would make it go to tableProperties?

File: spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/actions/BaseMigrateTableSparkAction.java
Comment (sririshindra) [reply spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/actions/BaseMigrateTableSparkAction.java]: I added the extra dropBackup method on MigrateTableSparkAction as you suggested in the latest commit. This way we avoid modifying the table properties.

File: api/src/main/java/org/apache/iceberg/actions/MigrateTable.java
Comment (szehon-ho) [inline api/src/main/java/org/apache/iceberg/actions/MigrateTable.java:45]: Capitlization looks a bit weird, let's revert Migration => migration?

File: api/src/main/java/org/apache/iceberg/actions/MigrateTable.java
Comment (szehon-ho) [inline api/src/main/java/org/apache/iceberg/actions/MigrateTable.java:47]: Can be more descriptive, maybe "if true, removes the backup table"

File: spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/actions/BaseMigrateTableSparkAction.java
Comment (szehon-ho) [inline spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/actions/BaseMigrateTableSparkAction.java:238]: Nit: grammar error here (should be- after + noun).  How about,  'Failed to drop the backup table {} after the migration'

File: .palantir/revapi.yml
Comment (szehon-ho) [inline .palantir/revapi.yml:6]: Why is partition key complaining here?  Can we not add this one?

Comment (sririshindra) [reply .palantir/revapi.yml:6]: I don't understand the reason why the partition key is complaining either. 
When I try to execute ` ./gradlew :iceberg-api:revapi --rerun-tasks`

I get the following error.

```
Execution failed for task ':iceberg-api:revapi'.
> There were Java public API/ABI breaks reported by revapi:
  
  java.class.defaultSerializationChanged: The default serialization ID for the class has changed. This means that the new version of the class is not deserializable from the byte stream of a serialized old class.
  
  old: class org.apache.iceberg.PartitionKey
  new: class org.apache.iceberg.PartitionKey
  
  SOURCE: EQUIVALENT, BINARY: EQUIVALENT, SEMANTIC: BREAKING
  
  From old archive: iceberg-api-0.14.0.jar
  From new archive: iceberg-api-0.15.0-SNAPSHOT.jar
```

File: spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/actions/BaseMigrateTableSparkAction.java
Comment (sririshindra) [reply spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/actions/BaseMigrateTableSparkAction.java]: I am not sure I understand the grammar error here, but I think it matches the style of the rest of the error messages in the file. But I changed `after migrate ` to `after migration`.

File: api/src/main/java/org/apache/iceberg/actions/MigrateTable.java
Comment (sririshindra) [reply api/src/main/java/org/apache/iceberg/actions/MigrateTable.java]: Done

File: api/src/main/java/org/apache/iceberg/actions/MigrateTable.java
Comment (sririshindra) [reply api/src/main/java/org/apache/iceberg/actions/MigrateTable.java]: Done

File: .palantir/revapi.yml
Comment (sririshindra) [reply .palantir/revapi.yml]: I didn't quite understand why this was needed. However, this was already added upstream, before this PR got merged. So, I think we can resolve this conversation. I updated the PR to fix the merge conflict.

**Comment by sririshindra** [comment]
@szehon-ho Please take another look at this PR when you get a chance. Thanks.

**Comment by sririshindra** [comment]
@szehon-ho Gentle ping to review this PR. Thanks.

File: api/src/main/java/org/apache/iceberg/actions/MigrateTable.java
Comment (szehon-ho) [inline api/src/main/java/org/apache/iceberg/actions/MigrateTable.java:45]: Sentence is a bit redudant, you can see DeleteOrphanFiles or ExpireSnapshots for more direct examples.  

The previous methods says 'sets a table property' because that's what it does, but here there's no point to say it sets a property.  How about:  'Drops the backup of the original table after a successful migration'.

File: api/src/main/java/org/apache/iceberg/actions/MigrateTable.java
Comment (szehon-ho) [inline api/src/main/java/org/apache/iceberg/actions/MigrateTable.java:47]: "if true, removes the backup table"

File: spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/actions/BaseMigrateTableSparkAction.java
Comment (szehon-ho) [inline spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/actions/BaseMigrateTableSparkAction.java:61]: Nit: we try to put a space between finals and non-finals as well

File: spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/actions/BaseMigrateTableSparkAction.java
Comment (szehon-ho) [reply spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/actions/BaseMigrateTableSparkAction.java]: Can you change the rest of the Spark files?   "after" should be used before a noun, and "migrate" is a verb.

**Review by szehon-ho — COMMENTED** [review]
Hi, was out for ApacheCon, code looks good, just some minor nits.

File: api/src/main/java/org/apache/iceberg/actions/MigrateTable.java
Comment (sririshindra) [reply api/src/main/java/org/apache/iceberg/actions/MigrateTable.java]: Done, Fixed in latest commit.

File: api/src/main/java/org/apache/iceberg/actions/MigrateTable.java
Comment (sririshindra) [reply api/src/main/java/org/apache/iceberg/actions/MigrateTable.java]: Done, Fixed in the latest commit.

File: spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/actions/BaseMigrateTableSparkAction.java
Comment (sririshindra) [reply spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/actions/BaseMigrateTableSparkAction.java]: I think the latest commit addresses the grammar issue correctly.  I used an online grammar checker to make sure the latest version is grammatically correct. Please let me know if a change is needed there. Thanks.

File: api/src/main/java/org/apache/iceberg/actions/MigrateTable.java
Comment (szehon-ho) [inline api/src/main/java/org/apache/iceberg/actions/MigrateTable.java:50]: @danielcweeks @rdblue wanted to check as its api module, we can add a new method if all subclass implement it, right?  Debating whether or not to make a default.

**Review by szehon-ho — APPROVED** [review]
Looks good to me.  Wanted to check if making this addition to api module is ok, should be to me but just to verify.

File: spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/actions/BaseMigrateTableSparkAction.java
Comment (szehon-ho) [reply spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/actions/BaseMigrateTableSparkAction.java]: Sounds good, it's a bit long but its fine.  Ref for the grammar comment (missed putting it earlier):  https://dictionary.cambridge.org/us/grammar/british-grammar/after-afterwards

File: api/src/main/java/org/apache/iceberg/actions/MigrateTable.java
Comment (szehon-ho) [reply api/src/main/java/org/apache/iceberg/actions/MigrateTable.java:50]: I checked with @danielcweeks offline, we could put a default that throws UnsupportedOperation or just leave it as is.  Probably the risk is small anyone implements it, so i think we are good.

File: spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/actions/BaseMigrateTableSparkAction.java
Comment (szehon-ho) [inline spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/actions/BaseMigrateTableSparkAction.java:102]: Can we rename the variables to be the same (either remove or drop).  It makes the code easier if there's fewer concepts, and we keep the same variable name.

**Review by szehon-ho — COMMENTED** [review]
I had a very small nit, sorry I did not notice it earlier, then we should be good to go.

File: spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/actions/BaseMigrateTableSparkAction.java
Comment (sririshindra) [reply spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/actions/BaseMigrateTableSparkAction.java]: Done. Fixed in the latest commit.

Comment (szehon-ho) [reply spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/actions/BaseMigrateTableSparkAction.java]: OK sorry, i guess it breaks the checkstyle :(.   

After taking a look through other Iceberg interfaces, I think nobody passes a boolean but just have a no-arg method.  Can we do that instead?  Thanks.

ie. 
dropBackup() {
   this.dropBackup = true;
   return this;
}

Comment (sririshindra) [reply spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/actions/BaseMigrateTableSparkAction.java]: I fixed the checkstyle issue in the latest commit. Thanks for your patience in reviewing this PR.

**Review by szehon-ho — APPROVED** [review]

**Comment by szehon-ho** [comment]
Thanks for the changes, will merge tomorrow if no further comments.

**Comment by szehon-ho** [comment]
Merged, thanks @sririshindra for the changes

**Pull request merged**
